### PR TITLE
new CLI

### PR DESCRIPTION
**Pull request by RoniSegal**

### CODE REVIEW

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 1]
```diff
@@ -10,12 +10,17 @@
...
 import io.sealights.plugins.sealightsjenkins.exceptions.SeaLightsIllegalStateException;
 import io.sealights.plugins.sealightsjenkins.utils.Logger;
+import io.sealights.plugins.sealightsjenkins.utils.PropertiesUtils;
+import io.sealights.plugins.sealightsjenkins.utils.StringUtils;
 import jenkins.model.Jenkins;
 import net.sf.json.JSONObject;
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.kohsuke.stapler.StaplerRequest;
 
+import java.awt.*;
```
Comment (shl3vi) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: is this dependency needed?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 2]
```diff
@@ -10,12 +10,17 @@
...
 import java.io.IOException;
+import java.util.*;
```
Comment (shl3vi) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: are this 2 dependencies needed?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 3]
```diff
@@ -33,6 +38,44 @@ public SealightsCLIBuildStep(boolean enabled, boolean failBuildIfStepFail,
...
+                this.commandMode instanceof CommandMode.StartView ||
+                this.commandMode instanceof CommandMode.ExternalReportView ||
+                this.commandMode instanceof CommandMode.UploadReportsView) {
+            if (!StringUtils.isNullOrEmpty(cliRunner.getAppName())) {
+                additionalArgs.append("appName=" + cliRunner.getAppName() + "\n");
+            }
+            if (!StringUtils.isNullOrEmpty(cliRunner.getBranchName())) {
+                additionalArgs.append("branchName=" + cliRunner.getBranchName() + "\n");
+            }
+
```
Comment (shl3vi) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: what about buildname? use 
`buildNameResolver.getFinalBuildName(build, envVars, buildName, logger))` to resolve the build name string

File: src/main/java/io/sealights/plugins/sealightsjenkins/utils/EnvVarsInjector.java [hunk 4]
```diff
@@ -42,7 +42,11 @@ private void printVariables() {
         logger.info("Injecting environment variables:");
         Set<String> keySet = additionalEnvVars.keySet();
         for (String key: keySet){
-            logger.info(key + " : " + additionalEnvVars.get(key));
+            if(key.equals("SL_TOKEN")){
```
Comment (shl3vi) [inline src/main/java/io/sealights/plugins/sealightsjenkins/utils/EnvVarsInjector.java]: try "SL_TOKEN".equals(key) -- that way you dodge case where key==null and cause NPE

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIRunner/config.jelly [hunk 5]
```diff
@@ -2,39 +2,39 @@
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout"
          xmlns:t="/lib/hudson" xmlns:f="/lib/form">
```
Comment (shl3vi) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIRunner/config.jelly]: Remove all the file we dont need it

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/EndView/help-additionalArguments.html [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                filesstorage
+            </td>
+            <td>
+                A folder where sealights files for this build will be kept (override the global 'Files storage' value
+                for this build).
+            </td>
+        </tr>
+        <tr>
+            <td>
+                appname
```
Comment (shl3vi) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/EndView/help-additionalArguments.html]: add descriptions to all new options
do this for all next 'help' files

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly [hunk 7]
```diff
@@ -16,14 +16,12 @@
                                  selected="${currentDescriptor == instance.commandMode.descriptor || (instance.commandMode.descriptor == null &amp;&amp; currentDescriptor.default)}"
                                  staplerClass="${currentDescriptor.clazz.name}">
                 <j:set var="instance" value="${instance.commandMode}"/>
-                <j:set var="descriptor" value="${instance.descriptor}"/>
```
Comment (shl3vi) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly]: maybe its worth adding documentation on the research and the results we discovered

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java [hunk 8]
```diff
@@ -19,9 +23,34 @@
...
+        this.buildSessionId = buildSessionId;
+    }
+
+    @Exported
+    public void setAdditionalArguments(String additionalArguments) {
+        this.additionalArguments = additionalArguments;
+    }
+
+    @Exported
+
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: Please remove the the empty line.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java [hunk 9]
```diff
@@ -229,8 +313,13 @@ public boolean isDefault() {
             public ConfigDescriptor() {
                 super(ConfigView.class, CommandModes.Config.getDisplayName());
             }
+
+            public DescriptorExtensionList<CommandBuildName, CommandBuildName.CommandBuildNameDescriptor> getBuildNameDescriptorList() {
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: Why do we need this? Please add comment about it.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 10]
```diff
@@ -33,6 +36,44 @@ public SealightsCLIBuildStep(boolean enabled, boolean failBuildIfStepFail,
...
+    private Object resolveFromCLiRunner() {
+        StringBuilder  additionalArgs = new StringBuilder();
+        if (this.commandMode instanceof CommandMode.ConfigView) {
+            if (!StringUtils.isNullOrEmpty(cliRunner.getAppName()))
+                ((CommandMode.ConfigView) commandMode).setAppName(cliRunner.getAppName());
+            if (!StringUtils.isNullOrEmpty(cliRunner.getBranchName()))
+                ((CommandMode.ConfigView) commandMode).setBranchName(cliRunner.getBranchName());
+            if (cliRunner.getBuildName() != null)
+                ((CommandMode.ConfigView) commandMode).setBuildName(cliRunner.getBuildName());
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: The code is a bit hard to read.
Maybe you can cast the object once and use it in all places?.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java (hunk 3 above)
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: +1

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 11]
```diff
@@ -33,6 +36,44 @@ public SealightsCLIBuildStep(boolean enabled, boolean failBuildIfStepFail,
...
+        }
+        return this;
+    }
+
+    private Object resolveFromCLiRunner() {
+        StringBuilder  additionalArgs = new StringBuilder();
+        if (this.commandMode instanceof CommandMode.ConfigView) {
+            if (!StringUtils.isNullOrEmpty(cliRunner.getAppName()))
+                ((CommandMode.ConfigView) commandMode).setAppName(cliRunner.getAppName());
+            if (!StringUtils.isNullOrEmpty(cliRunner.getBranchName()))
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: Maybe you should extract the code for ConfigView into one function and the code which handles the rest to another (BaseView)?

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/ConfigView/config.jelly [hunk 12]
```diff
@@ -12,4 +12,31 @@
...
+    </f:entry>
+
+    <f:entry title="Branch Name" field="branchName">
+        <f:textbox value="${branchName}" default="$${GIT_BRANCH}"/>
+    </f:entry>
+
+    <f:dropdownList name="buildName" title="Build Name">
+        <j:forEach var="currentDescriptor" items="${descriptor.buildNameDescriptorList}" varStatus="loop">
+            <f:dropdownListBlock title="${currentDescriptor.displayName}" value="${loop.index}"
+                                 selected="${currentDescriptor == instance.buildName.descriptor || (instance.buildName.descriptor == null &amp;&amp; currentDescriptor.emptyBuild)}"
```
Comment (nadavye) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/ConfigView/config.jelly]: Please ellaborate on this one.

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/ConfigView/config.jelly [hunk 13]
```diff
@@ -12,4 +12,31 @@
...
+                <j:set var="descriptor" value="${instance.descriptor}"/>
+                <st:include page="${currentDescriptor.configPage}" from="${currentDescriptor}"/>
+            </f:dropdownListBlock>
+        </j:forEach>
+    </f:dropdownList>
+    <f:advanced title="SeaLights CLI Advanced">
+        <f:entry title="Build Session Id" field="buildSessionId">
+            <f:textbox value="${buildSessionId}" />
+        </f:entry>
+        <f:entry title="${%Additional Arguments}" field="additionalArguments">
```
Comment (nadavye) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/ConfigView/config.jelly]: Why do we have a Precent sign?

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly [hunk 14]
```diff
@@ -10,20 +10,27 @@
         <f:checkbox checked="${it.failBuildIfStepFail}"/>
     </f:entry>
 
+    <!--
+        In jelly context we have few vars.
+        1. instance - an object currently being configured, within a section of a configure page such as a BuildStep; null if this is a newly added instance rather than reconfiguring.
+        2. descriptor - the Descriptor object (see below) corresponding to the class of instance.
+        this loop has two set statements:
+        the first set the instance of selected option to correct command mode (start, end, config, ..) otherwise its null.
+        the second set the descriptor of selected option to correct descriptor.
```
Comment (nadavye) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly]: Fix the 'otherwise is null'

Comment (nadavye) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly]: Quote  the 'set' word.

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/ConfigView/config.jelly (hunk 13 above)
Comment (RoniSegal) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/ConfigView/config.jelly]: from some reason without it the UI is blown

File: src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java [hunk 15]
```diff
@@ -572,43 +572,24 @@ private String joinPaths(String path1, String path2) {
...
-        return insertedBuildName;
-    }
-
-    private String getFinalBuildName(AbstractBuild<?, ?> build, SeaLightsPluginInfo slInfo, Logger logger) throws IllegalStateException {
-
+    
+    private String getFinalBuildName(AbstractBuild<?, ?> build, Properties additionalProps, Logger logger) throws IllegalStateException {
         String finalBuildName = null;
+        if (additionalProps.getProperty("buildname") != null) {
+            finalBuildName = additionalProps.getProperty("buildname").toString();
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java]: Why do you need to call 'toStirng()'? Seems like 'getProperty()' is returning string.

Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java]: Please fix everywhere.

File: src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java [hunk 16]
```diff
@@ -626,25 +607,29 @@ private SeaLightsPluginInfo createSeaLightsPluginInfo(
...
         slInfo.setRecursive(recursive);
-        slInfo.setPackagesIncluded(packagesIncluded);
-        slInfo.setPackagesExcluded(packagesExcluded);
+        slInfo.setPackagesIncluded(additionalProps.get("packagesincluded") != null ?
+                additionalProps.get("packagesincluded").toString() : null);
+        slInfo.setPackagesExcluded(additionalProps.get("packagesexcluded") != null ?
+                additionalProps.get("packagesexcluded").toString() : null);
         slInfo.setClassLoadersExcluded(classLoadersExcluded);
-        slInfo.setListenerJar(testListenerJar);
+        slInfo.setListenerJar((String) additionalProps.getProperty("testlistenerjar"));
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java]: Seems like this casting ("**(String)**additionalProps.getProperty('x');" is not needed

File: src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java [hunk 17]
```diff
@@ -741,12 +718,6 @@ private void setGlobalConfiguration(Logger logger, SeaLightsPluginInfo slInfo, P
 
         // set proxy
         String proxy = (String) additionalProps.get("proxy");
-        if (StringUtils.isNullOrEmpty(proxy)) {
-            proxy = override_proxy;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java]: Why did you remove this code?
Did you move the 'proxy' field to the "additional agruments"?

File: src/main/java/io/sealights/plugins/sealightsjenkins/MavenSealightsBuildStep.java [hunk 18]
```diff
@@ -114,44 +115,85 @@
...
+    public boolean isShowBackwardMessage(){
+        return !StringUtils.isBlank(backwardMessage);
+    }
+
     /**
      * The goal of this method is to support migration of data between versions
      * of this plugin.
      */
     private Object readResolve() {
+        return  beginAnalysis != null ? resolveFromBeginAnalysis() : this;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/MavenSealightsBuildStep.java]: This code is not clear. Please refactor or add comment.

File: src/main/java/io/sealights/plugins/sealightsjenkins/MavenSealightsBuildStep.java [hunk 19]
```diff
@@ -114,44 +115,85 @@
...
-            if (StringUtils.isNotBlank(targets)) {
-                if (enableSeaLights)
-                    this.buildStepMode = new BuildStepMode.InvokeMavenCommandView(targets);
-                else
-                    this.buildStepMode = new BuildStepMode.DisableSealightsView(targets);
-            } else {
-                //Assuming that users of 'MavenSealightsBuildStep' didn't left 'targets' blank.
-                //If no 'targets' is mentioned, its probably the deprecated 'beginAnalysisBuildStep'.
+    private Object resolveFromBeginAnalysis(){
+        setBackwardMessage("NOTICE: " );
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/MavenSealightsBuildStep.java]: Please check that 'beginAnalysis' isn't null.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/BuildStatusNotifier.java [hunk 20]
```diff
@@ -38,6 +39,7 @@
     private String branchName;
     private CommandBuildName buildName;
     private String additionalArguments;
+    private String reportTitle = "repot";
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/BuildStatusNotifier.java]: I assume 'repot' should be 'repo**r**t'?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/BuildStatusNotifier.java [hunk 21]
```diff
@@ -210,6 +211,21 @@ private Map createReportMap(Result result) {
...
+        Map map = new HashMap();
+        map.put("fieldName", fieldName);
+        map.put("type", type);
+        map.put("value", value);
+        return map;
+    }
+
+    private String toStringDuration(long duration){
+        long hours = TimeUnit.MILLISECONDS.toHours(duration) % 24;
+        long minutes = TimeUnit.MILLISECONDS.toMinutes(duration) % 60;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/BuildStatusNotifier.java]: A. Seems a bit risky. What happens if thats 36 hours? How will you represent it?
B. Please move it to another class which will handle this formatting as it is not part of the BuildStatusNotifier.java.
C. Add tests.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java [hunk 22]
```diff
@@ -80,7 +112,7 @@ public void setTestStage(String testStage) {
 
             @Override
             public boolean isDefault() {
-                return true;
+                return false;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: Why did you change it? Why 'false' and not 'true'?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java [hunk 23]
```diff
@@ -114,11 +146,12 @@ public EndDescriptor() {
...
         @DataBoundConstructor
-        public UploadReportsView(String reportFiles, String reportsFolders, boolean hasMoreRequests, String source) {
-            super(CommandModes.UploadReports);
+        public UploadReportsView(String reportFiles, String reportsFolders,
+                                 String source, String buildSessionId, String additionalArguments) {
+            super(CommandModes.UploadReports, buildSessionId, additionalArguments);
             this.reportFiles = reportFiles;
             this.reportsFolders = reportsFolders;
-            this.hasMoreRequests = hasMoreRequests;
+            this.hasMoreRequests = true;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: A. Is there a reason why the 'hasMoreRequests' is now 'true'?
B. If so, does anyone can override it? If not, can we remove the field?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java [hunk 24]
```diff
@@ -192,20 +225,89 @@ public ExternalReportDescriptor() {
...
-            this.packagesExcluded = packagesExcluded;
+        public ConfigView(String appName, String branchName,
+                          CommandBuildName buildName, String labId, String buildSessionId, String additionalArguments,
+                          List<TechnologyOptions> techOptions) {
+            super(CommandModes.Config, buildSessionId, additionalArguments);
+            this.appName = appName;
+            this.branchName = branchName;
+            this.buildName = buildName;
+            this.labId = labId;
+            this.techOptions = new DescribableList<>(Saveable.NOOP, Util.fixNull(techOptions));
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: Please explain this. This code is not clear.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 25]
```diff
@@ -33,6 +42,66 @@ public SealightsCLIBuildStep(boolean enabled, boolean failBuildIfStepFail,
...
+            _this = resolveFromCLiRunner();
+        }
+        if(commandMode instanceof CommandMode.ConfigView){
+            _this =resolveTechOption();
+        }
+        return _this;
+    }
+
+    private SealightsCLIBuildStep resolveTechOption(){
+        String packagesIncluded = ((CommandMode.ConfigView) commandMode).getPackagesIncluded();
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: Cast the CommanMode only once - no need to repeat '((CommandMode.ConfigView) commandMode)' over and over.

Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: You'll see that by deleting it, it will be easier to view the actual logic.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 26]
```diff
@@ -33,6 +42,66 @@ public SealightsCLIBuildStep(boolean enabled, boolean failBuildIfStepFail,
...
+        DescribableList<TechnologyOptions, TechnologyOptionsDescriptor> technologyOptions = new DescribableList<>(JavaOptions.DescriptorImpl.NOOP);
+        technologyOptions.add(new JavaOptions(packagesIncluded,packagesExcluded));
+        ((CommandMode.ConfigView) commandMode).setTechOptions(technologyOptions);
+        return this;
+    }
+
+    private SealightsCLIBuildStep resolveFromCLiRunner() {
+        StringBuilder  additionalArgs = new StringBuilder();
+        if (this.commandMode instanceof CommandMode.ConfigView) {
+               commandMode =setConfigparams();
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: A method which is setting something, shouldn't return a value. So, I assume that 'setConfigparams' should be called 'getConfigParams'.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 27]
```diff
@@ -33,6 +42,66 @@ public SealightsCLIBuildStep(boolean enabled, boolean failBuildIfStepFail,
         this.cliRunner = cliRunner;
     }
 
+    /* * The goal of this method is to support migration of data between versions
+    * of this plugin.
+    */
+    private Object readResolve() {
+        SealightsCLIBuildStep _this = this;
+        if (cliRunner != null) {
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: This code is not clear.
Why do you return 'this'? Seems like 'resolveTechOption()' and 'resolveFromCLiRunner()' return this, so there is no need to override '_this' with the result of those methods.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 28]
```diff
@@ -123,4 +194,31 @@ public String getDisplayName() {
...
     }
+
+    private CLIRunner createCLIRunner(CommandMode commandMode){
+        if(commandMode instanceof CommandMode.ConfigView){
+            CommandMode.ConfigView configView=(CommandMode.ConfigView) commandMode;
+            return new CLIRunner(configView.getBuildSessionId(),configView.getAppName(),configView.getBranchName(),
+                    configView.getBuildName(),configView.getAdditionalArguments(),null);
+        }else {
+            Properties properties = PropertiesUtils.toProperties(commandMode.getAdditionalArguments());
+            String appName = (properties.get("appname")!= null)? properties.get("appname").toString():null;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: No need to call 'toString()'. In addition, please check how proprerties.get() behaves when the key is not present.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/configurationtechnologies/JavaOptions.java [hunk 29]
```diff
@@ -0,0 +1,41 @@
+package io.sealights.plugins.sealightsjenkins.buildsteps.cli.configurationtechnologies;
+
+import hudson.Extension;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class JavaOptions extends TechnologyOptions{
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/configurationtechnologies/JavaOptions.java]: Please add documentation and explain the main goal of this class (Single Role Of Responsibility).

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/configurationtechnologies/DotNetOptions.java [hunk 30]
```diff
@@ -0,0 +1,45 @@
+package io.sealights.plugins.sealightsjenkins.buildsteps.cli.configurationtechnologies;
+
+import hudson.Extension;
+import hudson.model.AbstractProject;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.export.Exported;
+
+public class DotNetOptions /*extends TechnologyOptions*/ {
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/configurationtechnologies/DotNetOptions.java]: Why Java 'extends TechnologyOptions' and you don't do that in here?
Please add documentation on the class.

File: src/main/java/io/sealights/plugins/sealightsjenkins/utils/JenkinsUtils.java [hunk 31]
```diff
@@ -96,4 +98,27 @@ public String getWorkspace(AbstractBuild<?, ?> build) {
...
+        try {
+            List<Job> items = Jenkins.getInstance().getItems(Job.class);
+            String currentJobName = resolveCurrentJobName(descriptor);
+            for (Job j : items) {
+                if (j.getName().equals(currentJobName)) {
+                    return j;
+                }
+            }
+        }catch (Exception e){
+
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/utils/JenkinsUtils.java]: Don't catch exception and sallow it. Either log it or throw as RuntimeException.

Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/utils/JenkinsUtils.java]: Especially in infrastructure.

File: src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java (hunk 17 above)
Comment (RoniSegal) [inline src/main/java/io/sealights/plugins/sealightsjenkins/BeginAnalysis.java]: yes

File: src/main/java/io/sealights/plugins/sealightsjenkins/MavenSealightsBuildStep.java (hunk 19 above)
Comment (RoniSegal) [inline src/main/java/io/sealights/plugins/sealightsjenkins/MavenSealightsBuildStep.java]: its allready checked in readResolve function

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java (hunk 23 above)
Comment (RoniSegal) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: its Jason request, should be always true

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java (hunk 27 above)
Comment (RoniSegal) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: the method readResolve must return "this" for backward compatibility.
if you have multiple setup to upgrade you should use both methods, and  you cant access to "this" inside if.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java (hunk 28 above)
Comment (RoniSegal) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: get return Object appName is String.
when key doesn't  exist the method returns null

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/configurationtechnologies/DotNetOptions.java (hunk 30 above)
Comment (RoniSegal) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/configurationtechnologies/DotNetOptions.java]: we don't use DotNet option right now so we comment out the inheritance so its wont show on ui

File: src/main/java/io/sealights/plugins/sealightsjenkins/utils/JenkinsUtils.java (hunk 31 above)
Comment (RoniSegal) [inline src/main/java/io/sealights/plugins/sealightsjenkins/utils/JenkinsUtils.java]: this code currently  not used we save the logic only because its was hard to find out how to get the job name
